### DISCUSSION TITLE
level height, screen x/y, camera x/y

### DISCUSSION THREAD

**Discussion by city41** [city41/smaghetti#95] [Ideas]
Going to dump findings here as I get them. To start, I am pretty positive that "screen y boundary" is a 16 bit value of the level's height in pixels.

Here is mushroom14,room0 (a sky high adventure), it's setting for screen y boundary is `ff 03`, which translates to `0x3ff` -> `1023 pixels` -> `63.9375 tiles`.

When I change it to `fe 03`, this is what happens

![image](https://user-images.githubusercontent.com/141159/122114785-b0165c80-cdf1-11eb-8498-afbedfddfb5e.png)

bottom is the initial value, `ff 03`, and top is `fe 03`. If you look at the pattern in the wood, you can see the screen has shifted down by one pixel

when setting it to `00 04`, the screen shifts up by one pixel

![image](https://user-images.githubusercontent.com/141159/122115015-f370cb00-cdf1-11eb-8bc1-c1c4b4ac5780.png)

**Comment by city41** [comment]
this is  a spreadsheet of level settings from all rooms from all e-reader levels

https://docs.google.com/spreadsheets/d/17H2ObLplKjXfYogNK5wYdjnUPAIc5IkqhfZW_5mdINk

"fixed screen center y" is interesting, as it's either `00 00`, `10 80` or `10 81` in official levels. By playing around with these bytes I am seeing:

* if the second byte is `7f` or lower, these two bytes seem to have no impact
* if the second byte is `80` or higher, it shoots the camera way up to the top of the level
* if the second byte is `80` or higher, then the first byte seems to control the x position of the camera

So far nothing about these bytes say "fixed screen center y" to me. I feel like the names that historically were given to these level settings bytes may have been guesses.

**Reply by city41** [reply]
camera min and max both seem to be related to when mario is travelling down. 

# camera min
camera min seems to be how far mario has to go downward before the camera will start following him. set it to a high value like `ff` and the camera will essentially stop following mario. he can go way down below the camera and not die.

# camera max
if this is anything other than `00`, it seems to lock the camera from ever going back down again, and instead if mario falls off the bottom of the screen, this value indicates how far he has to fall before he dies. so with the tall level flying leap example, if this is `00`, mario can fall down the entire level. if it is `01` or higher, the camera will not follow mario, and instead him going off the bottom of the screen results in death.

worth noting none of the e-reader levels set camera max to anything other than `00`. I could see setting it for say levels that have rising lava, or auto scroll vertically. So maybe this is part of the game engine that smb3 doesn't need, but the engine is more versatile? I haven't played smb3 in a while, I don't remember any levels that cause mario to die if he falls off the bottom of the camera, but they are quite common in NSMB.

**Comment by city41** [comment]
you can now set the height of a room in smaghetti. But when you change it from the default, a warning is shown that changing height is experimental.

That's because changing height still causes strange things that need to be figured out. One of the biggest ones is in taller rooms (say 5+ screens high), it seems high up objects either get ignored or get placed more than once.

**Reply by city41** [reply]
the gba bg layer in which the level objects reside is 512x512

![image](https://user-images.githubusercontent.com/141159/122250329-92033780-ce97-11eb-99ca-4e3e598d4840.png)


so that explains why objects seem to get placed more than once in tall levels, as the map is just wrapping.

512px tall means 32 tiles tall, but A Sky-High Adventure's main room is 64 tiles tall. When you climb up to the very top of the level, then jump and go to the very bottom, the map updates as mario falls. So it seems there is something in a level that can enable it to be taller than 512px.

**Comment by city41** [comment]
I think in the end, this will be possible :crossed_fingers: 

![IMG_20210616_113705050](https://user-images.githubusercontent.com/141159/122262262-3b035f80-cea3-11eb-9e6e-1c685ba2cbc6.jpg)

**Comment by city41** [comment]
Smaghetti now allows room height to be set, capped off at a max of 63 tiles tall. It seems a sky high adventure does nothing special to handle the map wrapping. Instead, as long as a room is 64 tiles or less, the engine handles it just fine. 65 and greater, causes issues as the wrapping gets in the way.

I chose 63 just to have a slight buffer.

So in the end, rooms can still be quite massive. Honestly way too big considering smb3 has no check points :( with large rooms it becomes pretty easy to hit the max level size that can fit in a save file, but that's a problem for another day.

here is the largest a room can be

![IMG_20210618_120752944](https://user-images.githubusercontent.com/141159/122589254-08d23900-d02e-11eb-9b11-1e6dcac148bf.jpg)

**Comment by city41** [comment]
ok the final mystery is moving mario. Currently it's possible to move him anywhere in the level. But the camera starts out not on him in the y direction, and it has to pan to him as the level starts. Once the pan finishes, it's all just fine. I believe `player y screen center` is the key to that, which ... makes sense. I haven't played with those values recently, but in the past I remember them being very unintuitive. But anyway, getting to the end of this tunnel :+1:

**Reply by nintyalex** [reply]
Actually, it looks like `screen y` and `screen x` may be the answer. Fiddling with `player y screen center `only garbled the graphics (which render properly once you leave that screen and come back), but moving `screen x` and `screen y` to near Mario's new position properly centered the camera on him on start.

**Reply by city41** [reply]
Yeah, screen x/y do play a role here. I forgot to come back in here with my latest findings.

smaghetti now lets you move mario, but when you do it warns you it's experimental. Almost no matter what, when he gets moved from the default spot, and you play the level, the camera has to pan to him at least a little bit.

screenx/y being based on mario's x/y is important and gets us like 90% there, it's just that weird panning bit. It's possible other values that I have chosen that make the default location (lower left of the level) work make it hard to get other locations to work and thus all the values may need a rethink.

Here is what smaghetti is currently doing: https://github.com/city41/smaghetti/blob/91f3f48ee686822022e0810a7d713cea6bff3659/src/levelData/createLevelData.ts#L208-L237

**Reply by city41** [reply]
I have no idea what "player y screen center" actually is, I don't think it has anything to do with player y or the screen center :) 

interesting things from official e-reader levels:

* camera max is always zero
* the second byte of "player y screen center" is always zero
* "fixed screen center" is very strange as noted above, and I think incorrectly named
* there is no strong correlation that I can find between player y and the first byte of "player y screen center" :shrug:

**Reply by nintyalex** [reply]
I think I've figured out what the first byte of  `player y screen center` does. It centers Mario on the y position of the screen in pixels; if it's at `00`, the screen will only move upwards if Mario is at the top of the screen (`10` is one tile down, `20` two tiles down, and so on...). This is why the screen pans when starting a level in the middle of a room's height.

**Reply by city41** [reply]
Oh nice! I'll see if I can use this to fix the panning, which would be awesome.

**Reply by city41** [reply]
hmmm, still not exactly sure what that byte is doing. I can't get it to behave in a predictable way.

But I did find that `camera min` seems key. I had it set to `0x38`, which is what classic 1-2 sets it at. But setting it at `0x40`, and setting `screen starting y` to `playerY - 4` got a perfect result: Mario is centered nicely, no weird initial panning.

When camera min was `0x38`, there would be 3.5 tiles below Mario

![image](https://user-images.githubusercontent.com/141159/123276759-76211680-d4d3-11eb-9846-a1ee1dad2fff.png)

which makes sense, that's exactly what 0x38 works out to in pixels. This seems to force a pan, as `screen starting y` is in tiles, not pixels, so there is no way for it to match that straggling half tile.

Setting it to 0x40 gets 3 tiles below Mario

![image](https://user-images.githubusercontent.com/141159/123276977-aff21d00-d4d3-11eb-8642-5a5817f91f07.png)


I was expecting 4, hmmm. But, having camera min be a multiple of 16 seems to help a lot.

**Reply by city41** [reply]
if `camera min` and `screen start y` are defined in relation to each other, no weird panning happens.

these all place mario where you'd expect, and cause no panning

* screen start y: playerY - 5, camera min: 0x50
*  screen start y: playerY - 4, camera min: 0x40
*  screen start y: playerY - 3, camera min: 0x30
*  screen start y: playerY - 2, camera min: 0x20

basically that pattern seems to hold up. But at least for now, I think I will stick with camera min at `0x40` as many e-reader levels use that value. Main goal now is to just allow Mario to be placed anywhere in the level without any weirdness. Really fine camera/position control can come later.

**Reply by city41** [reply]
woohoo! this seems to work, I can't find a scenario where it doesn't. I think I'll move ahead with this and unmark moving mario as experimental.

**Reply by nintyalex** [reply]
Just one more thing... `player y screen center` must be less than or equal to `camera min` to prevent any panning issues as well.

**Reply by city41** [reply]
Ah ok. I'm still a little shaky on what it is doing. But I'm also a little shaky on what camera min does when it gets beyond 0x40-ish. I have a feeling these two will play a role in say allowing mario to start out at the very top of the viewport.

**Reply by nintyalex** [reply]
You're right. I tested your formula on those higher values (up to and including screen start y: playerY - playerY, camera min: 0x00), and they all worked perfectly.

**Reply by city41** [reply]
choosing `playerY - 4` and `0x40` for screen start y and camera min is what caused this bug: https://github.com/city41/smaghetti/issues/122

I've now fixed it by going with `playerY - 7` and `0x70`

Which is also what classic 1-2 uses, so in general sticking with what Nintendo did is usually safer :)

**Comment by city41** [comment]
A downside to the current camera solution is it really biases to above Mario. Recent levels that have been made end up where you can't see things just below Mario.

This was a known compromise, and I think the real solution is giving the user the ability to choose how the camera frames Mario. But since so many recent levels are affected by this, might bump it's priority up and tackle it sooner
